src/python_testing/TC_TSTAT_4_2.py

```diff
@@ -377,67 +474,106 @@ async def test_TC_TSTAT_4_2(self):
+                test_presets.append(new_preset)
+                presets_without_name_support = [new_preset]
 
-            # Clear state for next test.
-            await self.send_atomic_request_rollback_command()
+            if len(availableScenarios) > 0:
+
+                # Write to the presets attribute after setting a name for preset with handle (b'\x01') that doesn't support names
+                presets_without_name_support[0].name = "Name"
+
+                # Send the AtomicRequest begin command
+                await self.send_atomic_request_begin_command()
+
+                await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ConstraintError)
+
+                # Clear state for next test.
+                await self.send_atomic_request_rollback_command()
+            else:
+                logger.info(
+                    "Couldn't run test step 12 since there was no available preset scenario without name support")
 
         self.step("13")
         if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
 
-            # Send the AtomicRequest begin command
-            await self.send_atomic_request_begin_command()
+            availableScenario = self.get_available_scenario(presetTypes=presetTypes, presetScenarioCounts=presetScenarioCounts)
 
-            # Write to the presets attribute with a new valid preset added
-            test_presets = copy.deepcopy(new_presets_with_handle)
-            test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=cluster.Enums.PresetScenarioEnum.kWake,
-                                name="Wake", coolingSetpoint=2800, heatingSetpoint=1800, builtIn=False))
+            if len(current_presets) > 0 and len(current_presets) < numberOfPresetsSupported and availableScenario is not None:
 
-            await self.write_presets(endpoint=endpoint, presets=test_presets)
+                # Write to the presets attribute with a new valid preset added
+                test_presets = copy.deepcopy(current_presets)
+                test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=availableScenario,
+                                    coolingSetpoint=coolSetpoint, heatingSetpoint=heatSetpoint, builtIn=False))
 
-            # Roll back
-            await self.send_atomic_request_rollback_command()
+                # Send the AtomicRequest begin command
+                await self.send_atomic_request_begin_command()
+
+                await self.write_presets(endpoint=endpoint, presets=test_presets)
 
-            # Send the AtomicRequest commit command and expect InvalidInState as the previous edit request was cancelled
-            await self.send_atomic_request_commit_command(expected_status=Status.InvalidInState)
+                # Roll back
```

Comment:
Followup: Shouldn't we check that the presets before rollback are test_presets and after rollback are whatever they were at the start of this test?